Public/Src/Engine/Scheduler/Scheduler.cs

```diff
@@ -3529,21 +3529,29 @@ private async Task<PipExecutionStep> ExecutePipStep(RunnablePip runnablePip)
             }
         }
 
-        private static void FlagSharedOpaqueOutputs(IPipExecutionEnvironment environment, ExecutionResult executionResult)
+        private void FlagSharedOpaqueOutputs(IPipExecutionEnvironment environment, ProcessRunnablePip process)
         {
+            // Select all declared output files
+            IEnumerable<(AbsolutePath path, bool isKnownToBeUnderSharedOpaque)> paths = process.Process
+                .FileOutputs
+                .Select(f => (path: f.Path, isKnownToBeUnderSharedOpaque: false));
+
             // The shared dynamic accesses can be null when the pip failed on preparation, in which case it didn't run at all, so there is
             // nothing to flag
-            if (executionResult.SharedDynamicDirectoryWriteAccesses != null)
+            if (process.ExecutionResult.SharedDynamicDirectoryWriteAccesses != null)
             {
                 // Directory outputs are reported only when the pip is successful. So we need to rely on the raw shared dynamic write accesses,
                 // since flagging also happens on failed pips
-                foreach (IReadOnlyCollection<AbsolutePath> writesPerSharedOpaque in executionResult.SharedDynamicDirectoryWriteAccesses.Values)
-                {
-                    foreach (AbsolutePath writeInPath in writesPerSharedOpaque)
-                    {
-                        SharedOpaqueOutputHelper.EnforceFileIsSharedOpaqueOutput(writeInPath.ToString(environment.Context.PathTable));
-                    }
-                }
+                var sharedOpaqueDirOutputs = process.ExecutionResult
+                    .SharedDynamicDirectoryWriteAccesses
+                    .SelectMany(kvp => kvp.Value);
+
+                paths = paths.Concat(sharedOpaqueDirOutputs.Select(path => (path, isKnownToBeUnderSharedOpaque: true)));
+            }
+
+            foreach (var tuple in paths)
+            {
+                MakeSharedOpaqueOutputIfNeeded(tuple.path, force: tuple.isKnownToBeUnderSharedOpaque);

```

Comment:
>force [](start = 59, length = 5)

Why do we need this `force` field? For those in `SharedDynamicDirectoryWriteAccesses`, why don't we simply mark them as shared opaque outputs? And only check the graph for those in `FileOutputs`.

Reply:
Sure, I could break it up and do it twice: once for `SharedDynamicDirectoryWriteAccesses` (which get marked immediately, without checking) and once for `FileOutputs` (which are first checked and them marked if needed)

I did it this way so that I can concat them all together and uniformly call `MakeSharedOpaqueOutputIfNeeded`.  I can change it to what you suggested though

Reply:
Removed `force`